Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,591 @@
+            }
+
+            self.connections[index].failedToConnect()
+            return self.connections[index].eventLoop
+        }
+
+        // MARK: Connection lifecycle events
+
+        /// Sets the connection with the given `connectionId` to the draining state.
+        /// - Returns: the `EventLoop` to create a new connection on if applicable
+        /// - Precondition: connection with given `connectionId` must be either `.active` or already in the `.draining` state
+        mutating func goAwayReceived(_ connectionID: Connection.ID) -> GoAwayContext {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("go away recieved for a connection that does not exists")
+            }
+            let eventLoop = self.connections[index].goAwayReceived()
+            return GoAwayContext(eventLoop: eventLoop)
+        }
+
+        /// Update the maximum number of concurrent streams for the given connection.
+        /// - Parameters:
+        ///   - connectionID: The connectionID for which we received new settings
+        ///   - newMaxStreams: new maximum concurrent streams
+        /// - Returns: index of the connection and new number of available streams in the `AvailableConnectionContext`
+        /// - Precondition: Connections must be in the `.active` or `.draining` state.
+        mutating func newHTTP2MaxConcurrentStreamsReceived(
+            _ connectionID: Connection.ID,
+            newMaxStreams: Int
+        ) -> (Int, AvailableConnectionContext) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to update the maximum number of concurrent streams for a connection that does not exists")
+            }
+            let availableStreams = self.connections[index].newMaxConcurrentStreams(newMaxStreams)
+            let context = AvailableConnectionContext(
+                availableStreams: availableStreams,
+                eventLoop: self.connections[index].eventLoop,
+                isIdle: self.connections[index].isIdle
+            )
+            return (index, context)
+        }
+
+        // MARK: Leasing and releasing
+
+        mutating func leaseStream(onPreferred eventLoop: EventLoop) -> Connection? {
+            guard let index = self.findAvailableConnection(onPreferred: eventLoop) else { return nil }
+            return self.leaseStreams(at: index, count: 1)
+        }
+
+        /// tries to find an available connection on the prefered `eventLoop`. If it can't find one with the given `eventLoop`, it returns the first available connection
+        private func findAvailableConnection(onPreferred eventLoop: EventLoop) -> Int? {
+            var availableConnection: Int?
+            for (index, connection) in self.connections.enumerated() {
```

Comment:
We should be somewhat cautious here: `enumerated` does not return an `(index, element)` tuple, it returns an `(offset, element)` tuple. We should probably write a version of this that does what you actually want, in case we ever change the underlying object.